pkg/agent/iptables/iptables.go

```diff
@@ -119,6 +146,58 @@ func (c *Client) SetupRules() error {
 			return err
 		}
 	}
+
+	signals.ExitFns.Add(func() error {
```

Comment:
considering:

1. k8s pods by nature are ephemeral. Application has the ability to re-establish service. From application's perspective, this is no different than a service backing pod goes down, the service connection is re-established to another service backing pod.
2. place nice to next CNI getting installed, it could be an new version of Antrea or a CNI from different vendor which are unaware (old) Antrea and iptable rules which can break their operation.

Reply:
Proper deletion / cleanup is a separate concern IMO. The dataplane should not break IMO when the antrea agent is stopped (as part of a planned upgrade for example), and I would say this concern takes precedence.

There is a need to handle cleanup with a separate mechanism any way, since there is a always a possibility that Antrea does not exit gracefully when deleted.

Reply:
true, cleanup is separate issue. Will comment out.

Reply:
I gave some thoughts on this, and I just don't feel comfortable to have lingering route table entries or iptable entries, because next CNI may be
1. an non-antrea CNI or an Antrea upgrade that is unaware of the route table and iptable entries being added.
2. the same Antrea CNI but operates on different mode that is unaware of the route table and iptable entries

Before encap mode support, route table and iptable more or less stay static between each agent start, so there is little motivation to clear up them. Now with encap mode supports, iptable and route table are different when encap mode are different between each restart. So it is prudent to cleanup. 

Hope this make sense,

Reply:
For non-antrea CNI, we plan to have a separate cleanup mechanism after Antrea uninstall. @antoninbas has a proposal (#307) on this.
For other upgrade/config change case, Antrea should handle the cleanup after restart, or before it can handle that we have to require the separate cleanup step.
Agreed with what others said, we should first make sure no traffic outage when Agent is down/restarted.

Reply:
My thinking is
between Antrea restart:
1. if Antrea upgrade, the new version may have changed, the route table and iptables entries it modifies may also change, 
2. if Antrea change to another encapsulation mode, the new incarnation does not know what iptable and route table entries has been changed in previous incarnation

Reply:
I don't believe this change is high risk, it makes changing mode on the fly possible

Reply:
to continue my thoughts,
fundamentally I think it is important for Antrea to play nice with next CNI installed, be that newer version of Antrea, Antrea running on different mode, or another CNI as suggested by  https://github.com/vmware-tanzu/antrea/pull/307

From workflow perspective, keep in mind, with new noEncap/hybrid mode, iptables and route entries are different from those of encap mode.  this change allows Antrea simply to be reapplied without any other work. Without it, at least for encapsulation mode change (or even further upgrade), user may need to manually or an out-of-band script (may behttps://github.com/vmware-tanzu/antrea/pull/307) to remove iptables and route table entries from last Antrea installation before reapply with new encap mode.

From traffic outage perspective,
-- if upgrade, I am guessing both ovs/agent will be restarted, so traffic will be interrupted regardless
-- if agent crashed, today we capture only sigterm, sigint, so crash should not cause entries being wiped out.
-- and once again, k8s is build-in fault tolerant, From application level, all that matters is if there is service outage, and this change will not cause service outage.

Hope this make sense, Su

Reply:
I see normally upgrading/changing mode relies on next round to clean up the previous configuration, instead of previous round cleaning up itself, as it can be guaranteed previous round is terminated gracefully, a round might crash in the middle of setting iptables and routes, so it has to do the clean up anyway (and antrea already does such cleanup at startup). 
A new round doesn't need to know what mode or configuration was used in previous rounds, it should always clean up all iptables, routes managed by antrea and set up desired ones, of course in-place update is better. For example, all iptables rules managed by antrea should be in its own chains, so we can delete previous ones and install new ones (iptables-restore can make it easier).

Reply:
A similar case is kube-proxy, it won't do any cleanup when exiting, but cleans up previous iptables rules and sets up new ones after starting regardless of how its implementation is changed.

Reply:
it is much harder to clean up at restart, new incarnation do not know what iptable chains is used by previous incarnation. Iptable-save/restore would help, but there are not there yet today. Same goes for routing tables, given that we can have multiple route tables. 

but it is clear we prefer cleaning up on restart, will change accordingly. And open another PR to address clean up at restart.

Reply:
done. route and iptables are reconciled at agent starting up,